group_vars/mash_servers

```diff
@@ -971,3 +973,54 @@ devture_woodpecker_ci_agent_config_agent_secret: "{{ devture_woodpecker_ci_serve
 # /woodpecker-ci-agent                                                 #
 #                                                                      #
 ########################################################################
+
+########################################################################
+#                                                                      #
+# hubsite                                                              #
+#                                                                      #
+########################################################################
+
+hubsite_enabled: false
+
+hubsite_service_list_auto: |
+  {{
+    ([{'name': 'Miniflux', 'url': 'https://' +  miniflux_hostname + miniflux_path_prefix, 'logo_location': '{{ role_path }}/assets/miniflux.png', 'description': 'An opinionated feed reader '}] if miniflux_enabled else [])
+    +
+    ([{'name': 'Uptime Kuma', 'url': 'https://' + uptime_kuma_hostname + uptime_kuma_path_prefix, 'logo_location': '{{ role_path }}/assets/uptime-kuma.png', 'description': 'Check if the status of services'}] if uptime_kuma_enabled else [])
+    +
+    ([{'name': 'Nextcloud', 'url': 'https://' + nextcloud_hostname + nextcloud_path_prefix, 'logo_location': '{{ role_path }}/assets/nextcloud.png', 'description': 'Sync your files & much more'}] if nextcloud_enabled else [])
+    +
+    ([{'name': 'Peertube', 'url': 'https://' + peertube_hostname + peertube_path_prefix, 'logo_location': '{{ role_path }}/assets/peertube.png', 'description': 'Watch videos '}] if peertube_enabled else [])
+    +
+    ([{'name': 'Vaultwarden', 'url': 'https://' + vaultwarden_hostname + vaultwarden_path_prefix, ' logo_location': '{{ role_path }}/assets/vaultwarden.png', 'description': 'Securly access your passwords'}] if vaultwarden_enabled else [])
+    +
+    ([{'name': 'Gitea', 'url': 'https://' + gitea_hostname + gitea_path_prefix, 'logo_location': '{{ role_path }}/assets/gitea.png', 'description': 'A git service'}] if gitea_enabled else [])
+  }}
```

Comment:
In matrix-docker-ansible-deploy, we try to automatically wire services together. For example, you enable ma1sd and Synapse starts using it automatically, etc.

In mash-playbook, we try to stay way from such automatic wiring, because:

- it's difficult to reason about - sometimes there are unexpected consequences
- this playbook aims to support many many services and alternatives. Trying to auto-wire them all together can become messy

So far, the only automagic wiring we're doing is with Traefik, Postgres and Redis, but not among services. For example:

- Woodpecker CI is not automatically wired to Gitea when both are enabled. Our documentation tells you how to wire them manually
- the various container registry roles (Docker Registry; Docker Registry Browser; Docker Registry Purger) are not automatically wired together when enabled. Our documentation tells you how to wire them manually

Now, for Hubsite, it's a little different I suppose. It's probably another exception to this rule.

The beauty of something like Hubsite is that it's automatically populated and you won't need to worry about it.

I see a few problems though:

- sometimes, you would not like to advertise a service and expose its URL publicly (like Vaultwarden, for security reasons; or something else, for privacy reasons). It'd be nice to be able to easily disble it.
- it's probably difficult (but not impossible) for the Hubsite role to keep up with all the services we support - mostly in terms of logos and having to update th Hubsite role itself for this. Perhaps if the logos are stored in this playbook, it would be better, because each service contributor can hook the new service Hubsite.. And also, the logo can be used for another purpose as well - making our `docs/supported-services.md` page prettier by showing logos. Not sure what the legality of storing various logo images in this repo is, however. Perhaps it's good to have some indication that these images are licensed differently, etc.
- it'd be nice to order the list alphabetically by default. Perhaps custom ordering support would be nice too. Maybe adding some `priority` field (like the systemd service manager role does) is a possibility.
- sometimes, you'd be maintaining multiple MASH servers - splitting your service hosting among them. You may wish to run a single Hubsite dashboard on one of these servers though. Because this server enables no or few services, the Hubsite dashboard would be empty. Enabling a service require now requires toggling things like `gitea_enabled: true`, which has other undesirable side-effects. You'd like to enable it for Hubsite, but make the playbook actually install it on this server. I believe we need to introduce new variables (likely in `group_vars/mash_servers`) like this:

```yaml
hubsite_service_gitea_enabled: "{{ gitea_enabled }}"
hubsite_service_gitea_name: Gitea
hubsite_service_gitea_url: "'https://{{ gitea_hostname }}{{ gitea_path_prefix }}"
hubsite_service_gitea_logo_location: "{{ role_path }}/assets/gitea.png" # TODO - adjust to store the logos in the mash-playbook itself
hubsite_service_gitea_description: "A git service"
hubsite_service_gitea_priority: 1000
```

and then inject them into `hubsite_service_list_auto` like this:

```yaml
hubsite_service_list_auto: |
  {{
     ([{'name': hubsite_service_gitea_name, 'url': hubsite_service_gitea_url, 'logo_location': hubsite_service_gitea_logo_location, 'description': hubsite_service_gitea_description, 'priority': hubsite_service_gitea_priority}] if hubsite_service_gitea_enabled else [])
  }}
```

This way, everything is configurable. Even on a server managed by mash-playbook which doesn't install Gitea, you can expose a Gitea service by using the `hubsite_service_gitea_*` variables. You can also easily customize the name, description, logo and priority.

It may make sense for some of these `hubsite_service_gitea_*` variables to live inside the Hubsite role though, because putting them all in group vars will make the file even larger. But then.. it's more difficult for new service contributors to add new services.. Hmmm..

Reply:
I think you summed up pretty well what i struggled with so far. My current thought process was:
* the role should display service X nicely even if it is not aware of X -> no fixed list of services, user/playbook developer provides list and description of services.
* logos in the same style are best done with fixed process -> provide user with pre-made logos but allow custom (that might not look that good). Therefore I am not that thrilled by the idea to keep the logos (exclusively) in this playbook. A possibility would be to keep them in the role but maybe use the roles logos also in the playbook via external link?
```md
# Nextcloud
![Nextcloud Logo](https://github.com/moan0s/hubsite/blob/main/assets/nextcloud.png)
```
Ofc that would make the playbook somewhat dependent on this role which is not ideal. On the other hand it solves legality issues as necessary copyright is taken care of in the role (luckily wikipedia helps a lot ;))
A missing logo would not be nice, but also not terrible, we currently show just the service name which I think looks ok
*  Regarding introduce new variables in `group_vars`: I think it's the best solution currently, get's rid of the side-effects and enables us to do special things like "hiding" vaultwarden, I'll add that

Reply:
I think I addressed this now.